File: backend/app/schemas/application.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.db.postgres.models import ApplicationStatus, InterviewStatus


class ApplicationCreate(BaseModel):
    job_id: str
    candidate_id: str
    resume_id: str | None = None


class ApplicationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    job_id: str
    candidate_id: str
    resume_id: str | None
    status: ApplicationStatus
    match_score: float | None
    match_rationale: str | None
    ai_summary: str | None
    created_at: datetime
    updated_at: datetime


class ShortlistEntry(BaseModel):
    """One row of the recruiter-ready ranked shortlist."""
    application_id: str
    candidate_id: str
    candidate_name: str
    match_score: float = Field(..., ge=0, le=100)
    match_rationale: str
    top_skills: list[str] = Field(default_factory=list)
    passed_guardrails: bool
    bias_flags: list[str] = Field(default_factory=list)
    recommendation: str


class InterviewQuestion(BaseModel):
    id: str
    question: str
    category: str  # e.g. "technical", "behavioral", "system_design"


class InterviewCreate(BaseModel):
    application_id: str
    interviewer_name: str | None = None
    scheduled_at: datetime | None = None


class InterviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    application_id: str
    status: InterviewStatus
    questions: list[dict]
    overall_score: float | None
    score_breakdown: dict
    ai_recommendation: str | None
    created_at: datetime


class InterviewResponseIn(BaseModel):
    question_id: str
    question: str
    answer: str


class EvaluationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    application_id: str
    fairness_score: float | None
    bias_flags: list[str]
    grounding_score: float | None
    passed_guardrails: bool
    final_recommendation: str | None
    created_at: datetime

File: frontend/types/application.ts
export type ApplicationStatus =
  | "submitted"
  | "shortlisted"
  | "interviewing"
  | "offered"
  | "rejected"
  | "withdrawn";

export interface Application {
  id: string;
  job_id: string;
  candidate_id: string;
  resume_id: string | null;
  status: ApplicationStatus;
  match_score: number | null;
  match_rationale: string | null;
  ai_summary: string | null;
  created_at: string;
  updated_at: string;
}

/** The recruiter-facing ranked shortlist row — the primary output of the shortlist pipeline. */
export interface ShortlistEntry {
  application_id: string;
  candidate_id: string;
  candidate_name: string;
  match_score: number;
  match_rationale: string;
  top_skills: string[];
  passed_guardrails: boolean;
  bias_flags: string[];
  recommendation: string;
}

export type InterviewStatus = "pending" | "scheduled" | "in_progress" | "completed" | "cancelled";

export interface InterviewQuestion {
  id: string;
  question: string;
  category: "technical" | "behavioral" | "system_design" | "problem_solving" | "culture_fit";
}

export interface ScoreBreakdown {
  technical: number;
  communication: number;
  problem_solving: number;
  confidence: number;
  leadership: number;
}

export interface InterviewResponseRecord {
  id: string;
  question: string;
  answer: string | null;
  score: number | null;
  feedback: string | null;
}

export interface Interview {
  id: string;
  application_id: string;
  status: InterviewStatus;
  questions: InterviewQuestion[];
  overall_score: number | null;
  score_breakdown: ScoreBreakdown | Record<string, never>;
  ai_recommendation: string | null;
  created_at: string;
  responses: InterviewResponseRecord[];
}

export interface InterviewResponsePayload {
  question_id: string;
  question: string;
  answer: string;
}

export interface Evaluation {
  id: string;
  application_id: string;
  fairness_score: number | null;
  bias_flags: string[];
  grounding_score: number | null;
  passed_guardrails: boolean;
  final_recommendation: string | null;
  created_at: string;
}